### ISSUE TITLE
MIRI test failed on scene::mesh::buffer::test::test_add_attribute

### ISSUE THREAD

**Issue by kazuk** [FyroxEngine/Fyrox#155]
I cloned this repository yesterday. and I run `cargo +nightly miri test`

miri detects error `error: Undefined Behavior: incorrect layout on deallocation: alloc112996 has size 228 and alignment 4, but gave size 228 and alignment 1`

 https://github.com/rg3dengine/rg3d/blob/master/src/scene/mesh/buffer.rs#L340

miri report bellow.

```
    --> /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:104:14
     |
104  |     unsafe { __rust_dealloc(ptr, layout.size(), layout.align()) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ incorrect layout on deallocation: alloc112996 has size 228 and alignment 4, but gave size 228 and alignment 1
     |
     = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
     = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information

     = note: inside `std::alloc::dealloc` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:104:14
     = note: inside `<std::alloc::Global as std::alloc::Allocator>::deallocate` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:239:22
     = note: inside `<alloc::raw_vec::RawVec<u8> as std::ops::Drop>::drop` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:521:22
     = note: inside `std::ptr::drop_in_place::<alloc::raw_vec::RawVec<u8>> - shim(Some(alloc::raw_vec::RawVec<u8>))` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:192:1
     = note: inside `std::ptr::drop_in_place::<std::vec::Vec<u8>> - shim(Some(std::vec::Vec<u8>))` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:192:1
note: inside `scene::mesh::buffer::VertexBufferRefMut::add_attribute::<rg3d_core::nalgebra::Matrix<f32, rg3d_core::nalgebra::Const<2_usize>, rg3d_core::nalgebra::Const<1_usize>, rg3d_core::nalgebra::ArrayStorage<f32, 2_usize, 1_usize>>>` at src/scene/mesh/buffer.rs:340:13
    --> src/scene/mesh/buffer.rs:340:13
     |
340  |             self.vertex_buffer.data = new_data;
     |             ^^^^^^^^^^^^^^^^^^^^^^^
note: inside `scene::mesh::buffer::test::test_add_attribute` at src/scene/mesh/buffer.rs:1204:9
    --> src/scene/mesh/buffer.rs:1204:9
     |
1204 | /         buffer
1205 | |             .modify()
1206 | |             .add_attribute(
1207 | |                 VertexAttributeDescriptor {
...    |
1214 | |                 fill,
1215 | |             )
     | |_____________^
note: inside closure at src/scene/mesh/buffer.rs:1198:5
    --> src/scene/mesh/buffer.rs:1198:5
     |
1198 | /     fn test_add_attribute() {
1199 | |         let mut buffer = create_test_buffer();
1200 | |
1201 | |         let fill = Vector2::new(0.25, 0.75);
...    |
1277 | |         );
1278 | |     }
     | |_____^
     = note: inside `<[closure@src/scene/mesh/buffer.rs:1198:5: 1278:6] as std::ops::FnOnce<()>>::call_once - shim` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
     = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
     = note: inside `test::__rust_begin_short_backtrace::<fn()>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:577:5
     = note: inside closure at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:568:30
     = note: inside `<[closure@test::run_test::{closure#2}] as std::ops::FnOnce<()>>::call_once - shim(vtable)` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
     = note: inside `<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send> as std::ops::FnOnce<()>>::call_once` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1575:9
     = note: inside `<std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>> as std::ops::FnOnce<()>>::call_once` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:347:9
     = note: inside `std::panicking::r#try::do_call::<std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>>, ()>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:401:40
     = note: inside `std::panicking::r#try::<(), std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>>>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:365:19
     = note: inside `std::panic::catch_unwind::<std::panic::AssertUnwindSafe<std::boxed::Box<dyn std::ops::FnOnce() + std::marker::Send>>, ()>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:434:14
     = note: inside `test::run_test_in_process` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:600:18
     = note: inside closure at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:492:39
     = note: inside `test::run_test::run_test_inner` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:530:13
     = note: inside `test::run_test` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:564:28
     = note: inside `test::run_tests::<[closure@test::run_tests_console::{closure#2}]>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:305:17
     = note: inside `test::run_tests_console` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/console.rs:290:5
     = note: inside `test::test_main` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:122:15
     = note: inside `test::test_main_static` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/test/src/lib.rs:141:5
     = note: inside `main`
     = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
     = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125:18
     = note: inside closure at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:49:18
     = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:259:13
     = note: inside `std::panicking::r#try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:401:40
     = note: inside `std::panicking::r#try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:365:19
     = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:434:14
     = note: inside `std::rt::lang_start_internal` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:34:21
     = note: inside `std::rt::lang_start::<()>` at /home/kazuk/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:48:5
     = note: this error originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by mrDIMAS** [comment]
This is confusing, what `add_attribute` does there is just add gaps between existing data and fill them with copies of `fill_value` which should be perfectly fine because of `Copy` trait bound.

**Comment by kazuk** [comment]
I read that code, and rust std lib documents .

IMHO, 

MIRI detects miss-match alignment.
because Original VertexBuffer.data allocated as Vec<T>, but dropped as Vec<u8> .

https://github.com/rg3dengine/rg3d/blob/master/src/scene/mesh/buffer.rs#L380

Vec::from_raw_parts calls here.

https://github.com/rg3dengine/rg3d/blob/master/src/scene/mesh/buffer.rs#L386

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts

> T needs to have the same size and alignment as what ptr was allocated with. (T having a less strict alignment is not sufficient, the alignment really needs to be equal to satisfy the dealloc requirement that memory must be allocated and deallocated with the same layout.)

But this Undefined Behavior depends allocator implementation. 
May be safe on default allocator.

I think, some option to this issue

1. close this , because DEFAULT allocator may be safe.
2. add `Allocator requirements` for this project, "require safety to miss-match alignment on alloc-dealloc"
3. add type parameter to VertexBuffer,and change VertexBuffer.data type to `Vec<T>`

**Comment by mrDIMAS** [comment]
Ah, you right, thank you for the clarification! I wonder, what if I just copy input `data`'s bytes to `Vec<u8>` storage? This will ensure that data of type `T` will be dropped with correct alignment, and the new data will be of `u8` type and also will have correct alignment 🤔

**Comment by kazuk** [comment]
> Ah, you right, thank you for the clarification! I wonder, what if I just copy input data's bytes to Vec<u8> storage?

Can copy? `no-copy` NOT required?

use Vec::copy_from_slice for just memcpy.

 https://doc.rust-lang.org/std/vec/struct.Vec.html#method.copy_from_slice

**Comment by kazuk** [comment]
MIRI test pass on this code patch.

https://github.com/rg3dengine/rg3d/blob/master/src/scene/mesh/buffer.rs#L386

original code.

```rust
        let bytes =
           unsafe { Vec::<u8>::from_raw_parts(data.as_mut_ptr() as *mut u8, length, capacity) };

        std::mem::forget(data);
```

changed as

```rust
        let byte_slice =
            unsafe { std::slice::from_raw_parts(data.as_mut_ptr() as *mut u8, length) };
        let mut bytes = Vec::with_capacity(capacity);
        bytes.resize(length, 0u8);
        bytes.copy_from_slice(byte_slice);
```

I think , too much additional execution cost on this change. please check impact on real world.

add memory allocation on `Vec::with_capacity`, zero filling on `bytes.resize(length,0u8)`, memcpy on `bytes.copy_from_slice(byte_slice)`

**Comment by kazuk** [comment]
> I think , too much additional execution cost on this change. please check impact on real world.
> add memory allocation on Vec::with_capacity, zero filling on bytes.resize(length,0u8), memcpy on bytes.copy_from_slice(byte_slice)

minimize execution cost impact on #157

**Comment by kazuk** [comment]
tests completed on my box with MIRI  :tada: 

Ignore is

Too slow or hung on MIRI :snail: :turtle: 
- utils::astar::test::astar_random_points 
- utils::uvgen::test::test_generate_uvs

Unsupported by MIRI
- utils::lightmap::test::test_generate_lightmap : because File Access

Failed under MIRI
- utils::behavior::test::test_behavior_save_load : because using file system.

**Comment by mrDIMAS** [comment]
Fixed in #475
